Deal with memory leak caused by string and token

yetingk opened the pull request:
I let token lie on stack instead of heap, since I think we don't need token live too long.

yetingk committed:
Deal with memory leak caused by Token, string

Before Modification:
==15301== definitely lost: 81,536 bytes in 3,557 blocks
==15301== indirectly lost: 75,680 bytes in 1,147 blocks
==15301== possibly lost: 0 bytes in 0 blocks
==15301== still reachable: 544 bytes in 20 blocks
==15301== suppressed: 0 bytes in 0 blocks

After modification:
==14594== definitely lost: 71,344 bytes in 1,568 blocks
==14594== indirectly lost: 17,064 bytes in 549 blocks
==14594== possibly lost: 0 bytes in 0 blocks
==14594== still reachable: 544 bytes in 20 blocks
==14594== suppressed: 0 bytes in 0 blocks

jserv commented:
The commit messages should append the memory leak comparisons between w/ and w/o this patch via valgrind.

yetingk commented:
The difference between this modification when Inputing this command, 

valgrind --leak-check=full ./mzcc < sample/nqueen.c > sample/nqueen.s

Before modification:                                                                                                                                                                                            

==15301== 10,032 (24 direct, 10,008 indirect) bytes in 1 blocks are definitely lost in loss record 2,017 of 2,017
==15301==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15301==    by 0x404ED5: make_list (list.h:25)
==15301==    by 0x408626: read_toplevels (parser.c:1076)
==15301==    by 0x409DE6: main (main.c:9)
==15301== 
==15301== LEAK SUMMARY:
==15301==    definitely lost: 81,536 bytes in 3,557 blocks
==15301==    indirectly lost: 75,680 bytes in 1,147 blocks
==15301==      possibly lost: 0 bytes in 0 blocks
==15301==    still reachable: 544 bytes in 20 blocks
==15301==         suppressed: 0 bytes in 0 blocks

==15301== ERROR SUMMARY: 1447 errors from 1447 contexts (suppressed: 0 from 0)

After modification:

==14594== 10,032 (24 direct, 10,008 indirect) bytes in 1 blocks are definitely lost in loss record 1,050 of 1,050
==14594==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14594==    by 0x4050AE: make_list (list.h:25)
==14594==    by 0x408A52: read_toplevels (parser.c:1072)
==14594==    by 0x40A209: main (main.c:9)
==14594== 
==14594== LEAK SUMMARY:
==14594==    definitely lost: 71,344 bytes in 1,568 blocks
==14594==    indirectly lost: 17,064 bytes in 549 blocks
==14594==      possibly lost: 0 bytes in 0 blocks
==14594==    still reachable: 544 bytes in 20 blocks
==14594==         suppressed: 0 bytes in 0 blocks

==15445== ERROR SUMMARY: 778 errors from 778 contexts (suppressed: 0 from 0)

jserv commented:
> valgrind --leak-check=full ./mzcc < sample/nqueen.c > sample/nqueen.s
> Before modification:
> ==15301== definitely lost: 81,536 bytes in 3,557 blocks
> ==15301== indirectly lost: 75,680 bytes in 1,147 blocks
> After modification:
> ==14594== definitely lost: 71,344 bytes in 1,568 blocks
> ==14594== indirectly lost: 17,064 bytes in 549 blocks

The commit message was not accurate. This patch reduced memory leaks, but it still exists somewhere. We have to figure out the exact source.

jserv commented on verbose.c:
```diff
@@ -232,7 +233,8 @@ char *token_to_string(Token *tok)
         string_appendf(s, "\"%s\"", get_strtok(tok));
         return get_cstring(s);
+    default:
+        error("internal error: unknown token type: %d", get_ttype(tok));
```
Move this statement out of `switch` scope.

fakepaper56 replied:
But if I mover this statement out of switch, the output by "make check": 
verbose.c:219:5: error: enumeration value ‘TTYPE_NULL’ not handled in switch [-Werror=switch]
     switch (ttype) {
     ^~~~~~

jserv replied:
I misinterpreted the context.

jserv commented on util.h:
```diff
@@ -26,15 +26,16 @@ static inline String *make_string(void)
 {
-    return s->body;
+    char *ret = s->body;
+    free(s);
```
I am not sure if it necessary to call `free` here.

jserv replied:
All getter functions should touch data fields or members without releasing resources accordingly.

yetingk committed:
Deal with memory leak for list iterator and string

Before modification:
==14594== definitely lost: 71,344 bytes in 1,568 blocks
==14594== indirectly lost: 17,064 bytes in 549 blocks
==14594== possibly lost: 0 bytes in 0 blocks
==14594== still reachable: 544 bytes in 20 blocks
==14594== suppressed: 0 bytes in 0 blocks

After modification:
==14332==    definitely lost: 65,944 bytes in 893 blocks
==14332==    indirectly lost: 17,064 bytes in 549 blocks
==14332==      possibly lost: 0 bytes in 0 blocks
==14332==    still reachable: 544 bytes in 20 blocks
==14332==         suppressed: 0 bytes in 0 blocks

yetingk commented:
I use the same trick, struct initialization, on list iterator and strings

jserv commented:
Use `git rebase -i` to rework each git commit message to contain the summary of Valgrind report.

jserv commented on codegen_x64.c:
```diff
@@ -765,8 +765,8 @@ static void emit_func_prologue(Ast *func)
     }
-    for (Iter *i = list_iter(func->localvars); !iter_end(i);) {
-        Ast *v = iter_next(i);
+    for (Iter i = list_iter(func->localvars); !iter_end(i);) {
```
By the way, I am thinking of the introduce for [for_each macro](https://lwn.net/Articles/336255/) used in Linux kernel to simplify the similar loops here. Let's do that in another pull request.

jserv commented on util.h:
```diff
@@ -13,28 +13,28 @@ typedef struct {
-    r->body[0] = '\0';
-    return r;
+    char *body = malloc(INIT_SIZE);
+    body[0] = '\0';
```
Replace `malloc` with `calloc` to ensure allocated memory filled with bytes of value zero.

yetingk committed:
Use list to record cstring and free them at final

Before modification:
==14332==    definitely lost: 65,944 bytes in 893 blocks
==14332==    indirectly lost: 17,064 bytes in 549 blocks
==14332==      possibly lost: 0 bytes in 0 blocks
==14332==    still reachable: 544 bytes in 20 blocks
==14332==         suppressed: 0 bytes in 0 blocks

After modification:
==13827==    definitely lost: 64,896 bytes in 762 blocks
==13827==    indirectly lost: 16,904 bytes in 529 blocks
==13827==      possibly lost: 0 bytes in 0 blocks
==13827==    still reachable: 16,872 bytes in 801 blocks
==13827==         suppressed: 0 bytes in 0 blocks

yetingk committed:
Replace malloc with calloc

jserv commented on list.h:
```diff
@@ -109,4 +109,17 @@ static inline int list_len(List *list)
+static inline void list_free(List *list)
+{
+    for (Iter i = list_iter(list); !iter_end(i);)
+        iter_next_and_free(&i);
```
Safely move function body of `iter_next_and_free` here.

yetingk committed:
Remove iter_next_and_free()
